### DISCUSSION TITLE
Question: Is it possible to have control over comments while having them in extras?

### DISCUSSION THREAD

**Discussion by MichaHoffmann** [tree-sitter/tree-sitter#2186] [Q&A] [answered]
Hi,

I maintain a tree-sitter grammar for terraform and got the following issue: https://github.com/MichaHoffmann/tree-sitter-hcl/issues/30. I wrote the grammar in a way that comments are part of the `extras` of the grammar, which makes it possible for them to appear anywhere, which is correct but i wonder if it is possible to have more control over where they appear in the syntax tree. I wonder if the issue is solvable, does anyone have an idea?

Thank you!

**Comment by ahelwer** [comment]
You'd have to make the body node start before the comment. Currently it looks like you have the body node start at baz. You could, for example, move the open/close curly braces inside the body definition.

**Comment by ahlinc** [comment, marked as answer]
> You could, for example, move the open/close curly braces inside the body definition.

It's not possible cause the body may be empty and dangling curles in the body would look weird instead of absent body node.

Technically for parser the situation is correct because the body node isn't started right after the open curly and that means there are no tokens that belong to the body.

@MichaHoffmann I've opened a PR https://github.com/MichaHoffmann/tree-sitter-hcl/pull/31 that demonstrates how to cheat Tree-sitter parser with a fake token.
<details><summary>Screenshot</summary>

![Screenshot from 2023-04-08 07-02-02](https://user-images.githubusercontent.com/14666676/230702306-c1ca63f2-33dc-4cdd-a2bd-4c7de754da3a.png)
</details>

**Reply by MichaHoffmann** [reply]
Hey, thank you for the effort of opening a PR to demonstrate, much appreciated!

**Reply by MichaHoffmann** [reply]
I have to admit that i dont fully understand how it works. Is it because now we always can parse a body containing at least `SHIM` making the "optional" in
```
      block: $ => seq(
        $.identifier,
        repeat(choice($.string_lit, $.identifier)),
        $.block_start,
        optional($.body),
        $.block_end,
      ),
```
unnecessary?

**Reply by ahlinc** [reply]
No, it's not unnecessary, in case of a block that contains just comments it's better to omit the body node and leave comments between curly braces.
